### ISSUE TITLE
extracting labels from kmeans object in the same was as sklearn

### ISSUE THREAD

**Issue by prajjwal1** [facebookresearch/faiss#1236]
Hi,
I've looked at the documentation, where two snippets have been provided. It didn't make some things clear to me. I wanted to know how to get labels similar to how you would get from kmeans implementation in sklearn.

For instance,
1. I can do `.labels_` on the object returned after fit to get labels indicating which cluster the element belongs to.
How can this be done with faiss?
2. How can centroid indices be obtained? 

I've looked elsewhere but didn't find what I am looking for. Any help would be appreciated.

**Comment by mdouze** [comment]
See
https://github.com/facebookresearch/faiss/wiki/Faiss-building-blocks:-clustering,-PCA,-quantization#assignment

**Comment by prajjwal1** [comment]
I have already checked this. 
```
D, I = kmeans.index.search(x, 1)
```
returns nearest centroid for each line vector. I'm not sure what line vector refers to here.
```
index = faiss.IndexFlatL2 (d)
index.add (x)
D, I = index.search (kmeans.centroids, N)
```
returns nearest N points to the computed centroids.

But still they don't directly answer my two questions. Can you please provide some detail ? The documentation didn't make things clear. It would be good if some examples were present.

**Comment by prajjwal1** [comment]
Found some temporary solution. `MiniBatchKMeans` from sklearn is significantly faster then KMeans. There is a tradeoff between computation time and accuracy though.

Centroid indices can be found by computing pairwise distance (in sklearn) but can be used with FAISS also (both output numpy array :))
```
closest_centroid_indices, _ = pairwise_distances_argmin_min(kmeans.cluster_centers_, cls_embeddings) # kmeans.centroids
```

Not sure how to obtain `labels_` like array from FAISS kmeans object though.

**Comment by diegovalenzuelaiturra** [comment]
This may work

```python3
# imports
import faiss
import numpy as np

# create some random vectors
total_number_of_embeddings = 10
embedding_dim = 8

x = np.random.rand(*(total_number_of_embeddings, embedding_dim)).astype(
    dtype=np.float32)

# kmeans
k = 3
d = x.shape[1]

kmeans = faiss.Kmeans(d=d, k=k)
kmeans.train(x=x)

# results
cluster_centers_ = kmeans.centroids
labels_ = kmeans.index.search(x=x, k=1)[1].reshape(-1).astype(dtype=np.int32)
centroid_indices = kmeans.index.search(x=kmeans.centroids,
                                       k=1)[1].astype(dtype=np.int32)
```

**Comment by mdouze** [comment]
No activity, closing.
